src/library.js

```diff
@@ -428,28 +428,47 @@ LibraryManager.library = {
     ___setErrNo(ERRNO_CODES.EINVAL);
     return -1;
   },
-  sbrk: function(bytes) {
+
+  // Implement a Linux-like 'memory area' for our 'process'.
+  // Changes the size of the memory area by |bytes|; returns the
+  // address of the previous top ('break') of the memory area
+  // We control the "dynamic" memory - DYNAMIC_BASE to DYNAMICTOP
+  sbrk__asm: true,
+  sbrk__sig: ['ii'],
+  sbrk__deps: ['__setErrNo'],
+  sbrk: function(increment) {
+    increment = increment|0;
+    var dynamicTop = 0;
+    increment = ((increment + 15) & -16)|0;
 #if USE_PTHREADS
-    if (ENVIRONMENT_IS_PTHREAD) return _emscripten_sync_run_in_main_thread_1({{{ cDefine('EM_PROXIED_SBRK') }}}, bytes);
+    dynamicTop = Atomics_add(HEAPU32, DYNAMICTOP_PTR>>2, increment);
+#else
+    dynamicTop = HEAPU32[DYNAMICTOP_PTR>>2]|0;
+    HEAPU32[DYNAMICTOP_PTR>>2] = dynamicTop + increment|0;
 #endif
-    // Implement a Linux-like 'memory area' for our 'process'.
-    // Changes the size of the memory area by |bytes|; returns the
-    // address of the previous top ('break') of the memory area
-    // We control the "dynamic" memory - DYNAMIC_BASE to DYNAMICTOP
-    var self = _sbrk;
-    if (!self.called) {
-      DYNAMICTOP = alignMemoryPage(DYNAMICTOP); // make sure we start out aligned
-      self.called = true;
-      assert(Runtime.dynamicAlloc);
-      self.alloc = Runtime.dynamicAlloc;
-      Runtime.dynamicAlloc = function() { abort('cannot dynamically allocate, sbrk now has control') };
-    }
-    var ret = DYNAMICTOP;
-    if (bytes != 0) {
-      var success = self.alloc(bytes);
-      if (!success) return -1 >>> 0; // sbrk failure code
+    if ((dynamicTop|0) > (TOTAL_MEMORY|0)) {
+      // Warning: when -s USE_PTHREADS=1, this code block is called in a multithreaded context,
+      // so enlargeMemory() will be set to abort. If enlarging with pthreads is enabled at some point,
+      // sbrk will need to get mutexes to ensure that incrementing the heap size above and this enlarge
+      // operation is performed as a single transaction.
+      if ((enlargeMemory()|0) == 0) {
+        ___setErrNo({{{ cDefine('ENOMEM') }}});
+        return -1;
+      }
     }
-    return ret;  // Previous break location.
+    return dynamicTop|0;
+  },
+
+  brk__asm: true,
+  brk__sig: ['ii'],
```

Comment:
if we are adding `brk`, let's add a small test for it? untested code might mean the first time it's silently used by a new codebase it breaks in hard to find ways. or, add a warning on it in `ASSERTIONS` mode (as e.g. we do for some syscalls).